crates/bevy_ecs/src/world/command_queue.rs

```diff
@@ -120,7 +120,32 @@ impl CommandQueue {
         // flush the previously queued entities
         world.flush_entities();
 
+        // Apply any commands in the world's internal queue first.
+        while !world.command_queue.is_empty() {
+            // Put our queue into the `swap_queue`
+            std::mem::swap(self, &mut world.swap_queue);
+            // Take the full queue from the world and replace it with the `swap_queue`
+            std::mem::swap(self, &mut world.command_queue);
+            // Apply the queue
+            self.apply_or_drop_queued(Some(world));
```

Comment:
Hmm have we considered the implications of commands applying CommandQueues inside of this "already swapped" context?

Quick pseudo code / inline memory representation:

```rust
world.swap_queue = SWAP[]
world.command_queue = WORLD[A]
self = LOCAL[C]

A {
   // applies a new command queue (other_queue = OTHER[E]), which queues new world command D
}

self.apply() {
    while !world.command_queue.is_empty() {
        world.swap_queue = LOCAL[C]
        self = WORLD[A, B]
        world.command_queue = SWAP[]
        self.apply_or_drop_queued() {
            Apply(A) {
                other_queue.apply() {
                    // no-op because SWAP is empty (the current value of world.command_queue)
                    while !world.command_queue.is_empty() {
                    }

                    Apply(E) {
                        // Applying E inserts D into world.command_queue, which is currently SWAP
                        world.command_queue = SWAP[D]
                    }
                    
                    while !world.command_queue.is_empty() {
                        world.swap_queue = OTHER[]
                        other = SWAP[D]
                        world.command_queue = LOCAL[C] 
                        other.apply_or_drop_queued() {
                            // the world.command_queue is now LOCAL[C] for this apply
                            // C may be applied here
                        }
                    }
                }
            }
        }
        // WORLD is now empty and "ready" to be the new SWAP queue
        world.swap_queue = WORLD[]
        self = LOCAL[] // now empty (see comment above)
        world.command_queue = SWAP[]
        self.apply_or_drop_queued() {
            // The code was written as if C would be applied here, but it was actually applied above a couple of nestings deep, and may have been _followed_ by other commands created during 
        }
    }
}
```

For the "world.command_queue is now `LOCAL[C]` for this apply" comment above, doesn't this have pretty serious order of operations implications?

Reply:
Yeah that's not great, in that case we can just have an empty command queue on the stack that we move into `swap_queue` while we apply commands rather than inserting `self`.

Reply:
Might also be worth considering a sprinkle of unsafe to remove the swapping entirely. Both to avoid the brain breaking "nested swaps across contexts" situation and to optimize out a bunch of "unnecessary" swaps. Also, if we're creating a new empty queue on the stack, that also means we're allocating a bunch of extra times (and throwing it out on each call) instead of reusing already allocated space.

Reply:
We would only be allocating in the extra queue in the nested swap scenario, as long as the command queue application isn't recursive you would only be using the main two queues.

How do you picture unsafe being used here? We can't have the memory backing the command queue still in the world in any way since we could always end up writing to that queue due to some later nested command application and invalidating the memory we are reading.

Reply:
Given that we can read CommandMeta onto the stack, which is used to read the Command onto the stack, it seems like in theory we could ensure that there are no live references to the CommandQueue byte array while the command is being executed. If true, then provided we always look up directly using Vec indexing, then appending to the queue during command execution should be safe?

We'd also need to ensure we don't hold &mut CommandQueue over the command execution boundary.

Reply:
Yeah that's fair, it's a bit more of an involved refactor but I'll take a pass at it.

Reply:
So this is tricky in the case that a command panics. Currently we have nice behaviour where we set the length of the vector before starting the loop to ensure it's not re-used. We can't do that if we want to allow appending to partially applied queues since the vector still needs it's state, so we need to have a cursor stored in the queue that we increment before hand to ensure we don't re-drop/apply the panic-ed command.

If we are okay with the additional `usize` in `CommandQueue` we should be good.

Reply:
Ok, this doesn't really work at all if we want MIRI to be happy. If we keep any pointers to the command queue and later someone uses the world to get a mutable reference to the command queue MIRI complains. I tried boxing the command queue, but it doesn't matter, it always violates stacked borrow rules.

@cart what's your suggestion here?

Reply:
Spent a couple hours investigating this because I was pretty sure it should be possible. The solution I found includes the following:

1. Don't hold the mutable World borrow the whole time. Only turn it into a borrow at the exact point that it is used.
2. Box the World's CommandQueue. This decouples the memory from being stored directly on World, which ensures that `bytes` and `cursor` aren't "on the stack" when World is borrowed.
3. The addition of `RawCommandQueue`, which contains the raw pointers to `queue` and `cursor`. In addition to making some things a bit cleaner, it also allows us to use it in Commands (which was reworked to use an enum containing either `RawCommandQueue` or `Deferred<CommandQueue>`). This allows us to construct Commands from DeferredWorld without directly holding a mutable borrow to the CommandQueue.  I'm not convinced we actually need Commands to have the enum / branching. We might be able to have two separate types (one for `RawCommandQueue` and one for `Deferred<CommandQueue>`).

The branch is here: <https://github.com/cart/bevy/tree/raw_command_queue>. Needs a quality pass, a review of which new functions need to be unsafe, safety docs, and some design thinking about how RawCommandQueue should relate to Commands. I don't love the branching stuff.

I've gotta move on to other things for now. Let me know what you think / feel free to use whatever you want from that branch here.

Reply:
Definitely feels like we can get rid of the branching, but transitioning to only storing pointers should get around the stacked borrow problems MIRI was complaining about so this is closer to a solution than I am at the moment. 

Will merge this in, let people take a look, and then take another pass at clean-up when I have a chance.